src/transform-sdk/js/main.cc

```diff
@@ -315,8 +315,399 @@ std::expected<qjs::value, qjs::exception> make_write_event(
+          })
+          .transform([&obj](std::monostate) { return std::move(obj); });
+    }
+
+    static std::expected<qjs::value, qjs::exception>
+    make_reference(JSContext* ctx, const redpanda::sr::reference& ref) {
+        auto obj = qjs::value::object(ctx);
+        return qjs::value::string(ctx, ref.name)
+          .and_then([&obj](const qjs::value& name) {
+              return obj.set_property("name", name);
+          })
+          .and_then([ctx, &ref](std::monostate) {
+              return qjs::value::string(ctx, ref.subject);
+          })
+          .and_then([&obj](const qjs::value& subj) {
+              return obj.set_property("subject", subj);
+          })
+          .and_then([ctx, &obj, &ref](std::monostate) {
+              return obj.set_property(
+                "version", qjs::value::integer(ctx, ref.version));
+          })
+          .transform([&obj](std::monostate) { return std::move(obj); });
+    }
+};
+
+qjs::class_factory<schema_registry_client>
+make_schema_registry_client_class(qjs::runtime* runtime) {
+    qjs::class_builder<schema_registry_client> builder(
+      runtime->context(), "SchemaRegistryClient");
+    builder.method<&schema_registry_client::lookup_schema_by_id>(
+      "lookupSchemaById");
+    builder.method<&schema_registry_client::lookup_schema_by_version>(
+      "lookupSchemaByVersion");
+    builder.method<&schema_registry_client::lookup_latest_schema>(
+      "lookupLatestSchema");
+    builder.method<&schema_registry_client::create_schema>("createSchema");
+    return builder.build();
+}
+
+redpanda::bytes_view value_as_bytes_view(const qjs::value& val) {
+    if (val.is_string()) {
+        auto data = val.string_data();
+        return redpanda::bytes_view{data.view()};
+    }
+    if (val.is_array_buffer()) {
+        auto data = val.array_buffer_data();
+        return redpanda::bytes_view{data.data(), data.size()};
+    }
+    if (val.is_uint8_array()) {
+        auto data = val.uint8_array_data();
+        return redpanda::bytes_view{data.data(), data.size()};
+    }
+    return redpanda::bytes_view{};
+}
+
+std::expected<qjs::value, qjs::exception>
+decode_schema_id_impl(JSContext* ctx, qjs::value& arg) {
+    auto data = value_as_bytes_view(arg);
+    auto obj = qjs::value::object(ctx);
+    return redpanda::sr::decode_schema_id(data)
+      .transform_error([ctx](std::error_code err) {
+          return qjs::exception::make(
+            ctx, std::format("Failed to decode schema ID: {}", err.message()));
+      })
+      .and_then(
+        [ctx,
+         &obj](std::pair<redpanda::sr::schema_id, redpanda::bytes_view> sid) {
+            return obj.set_property("id", qjs::value::integer(ctx, sid.first));
+        })
+      .and_then(
+        [ctx,
+         &arg](std::monostate) -> std::expected<qjs::value, qjs::exception> {
+            // NOTE(oren): returning a view back into the target buffer might
+            // be a terrible idea. Needs a good think.
```

Comment:
Let's make a copy for now and optimize later. We should be able to keep a reference to the old data and make a subview: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/subarray

Reply:
done

Reply:
done